tests/test_syntax/blocks/test_html_blocks.py

```diff
@@ -0,0 +1,783 @@
+            ),
+            self.dedent(
+                """
+                <?php
+                echo '>';'
+                ?>
+                """
+            )
+        )
+
+    def test_raw_processing_instruction_with_blank_lines(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <?php
+
+                echo '>';'
+
+                ?>
+                """
+            ),
+            self.dedent(
+                """
+                <?php
+
+                echo '>';'
+
+                ?>
+                """
+            )
+        )
+
+    def test_raw_processing_instruction_indented(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <?php
+
+                    echo '>';'
+
+                ?>
+                """
+            ),
+            self.dedent(
+                """
+                <?php
+
+                    echo '>';'
+
+                ?>
+                """
+            )
+        )
+
+    def test_raw_declaration_one_line(self):
+        self.assertMarkdownRenders(
+            '<!DOCTYPE html>',
+            '<!DOCTYPE html>'
+        )
+
+    # TODO: Decide correct behavior. This matches current behavior and Commonmark.
+    # The reference implementation considers this inline not block level:
+    #
+    #   <p><!DOCTYPE html><em>bar</em></p>
+    #
+    # But most implementations do this instead:
+    #
+    #   <p>&lt;!DOCTYPE html&gt;<em>bar</em></p>
+    #
+    # Either makes sense, but the later seems more correct to me.
+    def test_raw_declaration_one_line_followed_by_text(self):
+        self.assertMarkdownRenders(
+            '<!DOCTYPE html>*bar*',
+            '<!DOCTYPE html>*bar*'
+        )
+
+    def test_raw_multiline_declaration(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <!DOCTYPE html PUBLIC
+                  "-//W3C//DTD XHTML 1.1//EN"
+                  "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
+                """
+            ),
+            self.dedent(
+                """
+                <!DOCTYPE html PUBLIC
+                  "-//W3C//DTD XHTML 1.1//EN"
+                  "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
+                """
+            )
+        )
+
+    def test_raw_cdata_one_line(self):
+        self.assertMarkdownRenders(
+            '<![CDATA[ document.write(">"); ]]>',
+            '<![CDATA[ document.write(">"); ]]>'
+        )
+
+    # TODO: Decide correct behavior. This matches current behavior and Commonmark.
+    # The reference implementation considers this inline not block level:
+    #
+    #   <p><![CDATA[ document.write(">"); ]]><em>bar</em></p>
+    #
+    # But most implementations do this instead:
+    #
+    #   <p>&lt;[CDATA[ document.write(“&gt;”); ]]&gt;<em>bar</em></p>
+    #
+    # Either makes sense, but the later seems more correct to me.
```

Comment:
Your suggestion here seems to differ from all of the other situations were plain text follows and otherwise block-level tag on the same line. Any reason why?

Reply:
Whether CDATA is valid inside the document depends on whether it is HTML or XHTML. As our default output format is HTML (and XHTML is generally considered deprecated nowadays), I think it is better to escape occurrences of CDATA in our output.

Reply:
Most browsers just ignore CDATA in HTML5. I would personally just treat it as a preprocessor and move on. I personally don't like the idea of treating CDATA as text content.

Reply:
@mitya57 actually our default output format is XHTML. As a reminder, the reference implementation outputs XHTML and the rules specifically expect XHTML output. We will not be changing our default as long as that is the case.

Of course, in practice that makes no difference for HTML5 as both XHTML and HTML style tags are valid in HTML5. And as we only output fragments, not entire documents, the tag style is the only difference any output format can effect.

In any event, if we were going to escape or not escape CDATA based on output format, then that should be determined in the serializer, not in the HTML parser. Personally, I'm inclined to not bother with that. It is the document authors responsibility to ensure his raw HTML is valid, not ours. If including CDATA makes for an invalid document, then it shouldn't be included by the document author. It is not our place to make that determination.

Reply:
> actually our default output format is XHTML.

Ah, indeed.

> Personally, I'm inclined to not bother with that. It is the document authors responsibility to ensure his raw HTML is valid, not ours.

Sounds fine.